src/datastructure/Timetable.cpp

```diff
@@ -1,212 +1,333 @@
+   *
+   * @return The index of the train in the train list.
+   */
+  if (!network.has_vertex(entry)) {
+    throw exceptions::VertexNotExistentException(entry);
+  }
+  if (!network.has_vertex(exit)) {
+    throw exceptions::VertexNotExistentException(exit);
+  }
+  if (train_list.has_train(name)) {
+    throw exceptions::ConsistencyException("Train already exists.");
+  }
+  auto const index =
+      train_list.add_train(name, length, max_speed, acceleration, deceleration);
+  schedules.emplace_back(t_0, v_0, entry, t_n, v_n, exit);
+  return index;
 }
 
-void cda_rail::Timetable::add_stop(int train_index, const std::string &station_name, int begin, int end, bool sort) {
-    if (!train_list.has_train(train_index)) {
-        throw std::invalid_argument("Train does not exist.");
-    }
-    if (!station_list.has_station(station_name)) {
-        throw std::invalid_argument("Station does not exist.");
-    }
-    if (begin < 0 || end < 0) {
-        throw std::invalid_argument("Time cannot be negative.");
-    }
-    if (begin >= end) {
-        throw std::invalid_argument("End time has to be after the start time.");
-    }
+void cda_rail::Timetable::add_stop(size_t             train_index,
+                                   const std::string& station_name, int begin,
+                                   int end, bool sort) {
+  /**
+   * This method adds a stop to a train schedule. The stop is specified by its
+   * parameters.
+   *
+   * @param train_index The index of the train in the train list.
+   * @param station_name The name of the station.
+   * @param begin The time at which the train stops at the station in s.
+   * @param end The time at which the train leaves the station in s.
+   * @param sort If true, the stops are sorted after insertion.
+   */
+  if (!train_list.has_train(train_index)) {
+    throw exceptions::TrainNotExistentException(train_index);
+  }
+  if (!station_list.has_station(station_name)) {
+    throw exceptions::StationNotExistentException(station_name);
+  }
+  if (begin < 0 || end < 0) {
+    throw exceptions::InvalidInputException("Time cannot be negative.");
+  }
+  if (begin >= end) {
+    throw exceptions::ConsistencyException(
+        "End time has to be after the start time.");
+  }
 
-    auto& stops_reference = schedules.at(train_index).stops;
-    for (const auto& stop : stops_reference) {
-        if (stop.station == station_name) {
-            throw std::invalid_argument("Train already stops at station.");
-        }
-        // Check if [begin, end] and [stop.begin, stop.end] overlap
-        if (begin <= stop.end && end >= stop.begin) {
-            throw std::invalid_argument("Train has another stop at this time.");
-        }
+  auto& stops_reference = schedules.at(train_index).stops;
+  for (const auto& stop : stops_reference) {
+    if (stop.station == station_name) {
+      throw exceptions::ConsistencyException("Train already stops at station.");
     }
 
```

Comment:
Just a thought. Could a train not stop at a station twice if it does like a round-trip? I guess this is a rare occurence though.

Reply:
I think allowing this would have many complicating consequences, hence, not allow round-trips.